File: learning-factory/warehouse/artifacts/projects/build-your-own-x/programming-language/creating-jvm-language-7cdaf40c49/job_byox_build_v1_5d3af01bcd55594749205d85322b9429/attempt-001/sealed/reference/src/main/java/dev/learningfactory/sprig/Parser.java
package dev.learningfactory.sprig;

import java.util.ArrayList;
import java.util.List;

final class Parser {
    private final List<Token> tokens;
    private int current;
    private int statementNesting;
    private int recursiveExpressionNesting;

    Parser(List<Token> tokens) {
        this.tokens = List.copyOf(tokens);
    }

    Ast.Program parse() throws CompileFailure {
        Token start = consume(TokenType.FN, "expected 'fn'");
        consume(TokenType.MAIN, "expected 'main'");
        consume(TokenType.LEFT_PAREN, "expected '('");
        consume(TokenType.RIGHT_PAREN, "expected ')'");
        consume(TokenType.ARROW, "expected '->'");
        consume(TokenType.INT, "expected return type 'Int'");
        List<Ast.Stmt> body = block();
        consume(TokenType.EOF, "unexpected trailing input");
        return new Ast.Program(body, start.pos());
    }

    private List<Ast.Stmt> block() throws CompileFailure {
        Token open = consume(TokenType.LEFT_BRACE, "expected '{'");
        statementNesting++;
        if (statementNesting > Ast.MAX_NESTING) {
            throw CompileFailure.at("E_LIMIT", open.line(), open.column(),
                    "statement nesting exceeds " + Ast.MAX_NESTING);
        }
        List<Ast.Stmt> statements = new ArrayList<>();
        while (!check(TokenType.RIGHT_BRACE) && !check(TokenType.EOF)) {
            statements.add(statement());
        }
        consume(TokenType.RIGHT_BRACE, "expected '}'");
        statementNesting--;
        return List.copyOf(statements);
    }

    private Ast.Stmt statement() throws CompileFailure {
        if (match(TokenType.LET)) {
            Token keyword = previous();
            Token name = consume(TokenType.IDENTIFIER, "expected variable name");
            consume(TokenType.ASSIGN, "expected '=' after variable name");
            Ast.Expr initializer = expression();
            consume(TokenType.SEMICOLON, "expected ';' after declaration");
            return new Ast.Let(name.text(), initializer, keyword.pos());
        }
        if (match(TokenType.PRINT)) {
            Token keyword = previous();
            Ast.Expr value = expression();
            consume(TokenType.SEMICOLON, "expected ';' after print value");
            return new Ast.Print(value, keyword.pos());
        }
        if (match(TokenType.IF)) {
            Token keyword = previous();
            consume(TokenType.LEFT_PAREN, "expected '(' after 'if'");
            Ast.Expr condition = expression();
            consume(TokenType.RIGHT_PAREN, "expected ')' after condition");
            List<Ast.Stmt> thenBranch = block();
            consume(TokenType.ELSE, "expected 'else'");
            List<Ast.Stmt> elseBranch = block();
            return new Ast.If(condition, thenBranch, elseBranch, keyword.pos());
        }
        if (match(TokenType.WHILE)) {
            Token keyword = previous();
            consume(TokenType.LEFT_PAREN, "expected '(' after 'while'");
            Ast.Expr condition = expression();
            consume(TokenType.RIGHT_PAREN, "expected ')' after condition");
            return new Ast.While(condition, block(), keyword.pos());
        }
        if (match(TokenType.RETURN)) {
            Token keyword = previous();
            Ast.Expr value = expression();
            consume(TokenType.SEMICOLON, "expected ';' after return value");
            return new Ast.Return(value, keyword.pos());
        }
        if (check(TokenType.IDENTIFIER)) {
            Token name = advance();
            consume(TokenType.ASSIGN, "expected '=' after variable name");
            Ast.Expr value = expression();
            consume(TokenType.SEMICOLON, "expected ';' after assignment");
            return new Ast.Assign(name.text(), value, name.pos());
        }
        throw syntax(peek(), "expected a statement");
    }

    private Ast.Expr expression() throws CompileFailure {
        return or();
    }

    private Ast.Expr or() throws CompileFailure {
        return leftAssociative(this::and, TokenType.OR_OR);
    }

    private Ast.Expr and() throws CompileFailure {
        return leftAssociative(this::equality, TokenType.AND_AND);
    }

    private Ast.Expr equality() throws CompileFailure {
        return leftAssociative(this::relation, TokenType.EQUAL_EQUAL,
                TokenType.BANG_EQUAL);
    }

    private Ast.Expr relation() throws CompileFailure {
        return leftAssociative(this::term, TokenType.LESS, TokenType.LESS_EQUAL,
                TokenType.GREATER, TokenType.GREATER_EQUAL);
    }

    private Ast.Expr term() throws CompileFailure {
        return leftAssociative(this::factor, TokenType.PLUS, TokenType.MINUS);
    }

    private Ast.Expr factor() throws CompileFailure {
        return leftAssociative(this::unary, TokenType.STAR, TokenType.SLASH,
                TokenType.PERCENT);
    }

    private Ast.Expr leftAssociative(ExpressionParser operand, TokenType... operators)
            throws CompileFailure {
        Ast.Expr expression = operand.parse();
        while (match(operators)) {
            Token operator = previous();
            Ast.Expr right = operand.parse();
            int depth = 1 + Math.max(expression.depth(), right.depth());
            if (depth > Ast.MAX_NESTING) {
                throw CompileFailure.at("E_LIMIT", operator.line(), operator.column(),
                        "expression nesting exceeds " + Ast.MAX_NESTING);
            }
            expression = new Ast.Binary(operator.type(), expression, right,
                    operator.pos(), depth);
        }
        return expression;
    }

    private Ast.Expr unary() throws CompileFailure {
        if (match(TokenType.BANG, TokenType.MINUS)) {
            Token operator = previous();
            enterExpression(operator);
            Ast.Expr operand;
            try {
                operand = unary();
            } finally {
                recursiveExpressionNesting--;
            }
            int depth = operand.depth() + 1;
            if (depth > Ast.MAX_NESTING) {
                throw CompileFailure.at("E_LIMIT", operator.line(), operator.column(),
                        "expression nesting exceeds " + Ast.MAX_NESTING);
            }
            return new Ast.Unary(operator.type(), operand, operator.pos(), depth);
        }
        return primary();
    }

    private Ast.Expr primary() throws CompileFailure {
        if (match(TokenType.INTEGER)) {
            Token token = previous();
            return new Ast.Literal(Integer.parseInt(token.text()), Ast.Type.INT, token.pos());
        }
        if (match(TokenType.TRUE)) {
            Token token = previous();
            return new Ast.Literal(1, Ast.Type.BOOL, token.pos());
        }
        if (match(TokenType.FALSE)) {
            Token token = previous();
            return new Ast.Literal(0, Ast.Type.BOOL, token.pos());
        }
        if (match(TokenType.IDENTIFIER)) {
            Token token = previous();
            return new Ast.Variable(token.text(), token.pos());
        }
        if (match(TokenType.LEFT_PAREN)) {
            Token open = previous();
            enterExpression(open);
            Ast.Expr nested;
            try {
                nested = expression();
                consume(TokenType.RIGHT_PAREN, "expected ')' after expression");
            } finally {
                recursiveExpressionNesting--;
            }
            return nested;
        }
        throw syntax(peek(), "expected an expression");
    }

    private void enterExpression(Token token) throws CompileFailure {
        recursiveExpressionNesting++;
        if (recursiveExpressionNesting > Ast.MAX_NESTING) {
            throw CompileFailure.at("E_LIMIT", token.line(), token.column(),
                    "expression nesting exceeds " + Ast.MAX_NESTING);
        }
    }

    private Token consume(TokenType type, String message) throws CompileFailure {
        if (check(type)) return advance();
        throw syntax(peek(), message);
    }

    private boolean match(TokenType... types) {
        for (TokenType type : types) {
            if (check(type)) {
                advance();
                return true;
            }
        }
        return false;
    }

    private boolean check(TokenType type) {
        return peek().type() == type;
    }

    private Token advance() {
        if (!check(TokenType.EOF)) current++;
        return previous();
    }

    private Token peek() {
        return tokens.get(current);
    }

    private Token previous() {
        return tokens.get(current - 1);
    }

    private static CompileFailure syntax(Token token, String message) {
        return CompileFailure.at("E_SYNTAX", token.line(), token.column(), message);
    }

    @FunctionalInterface
    private interface ExpressionParser {
        Ast.Expr parse() throws CompileFailure;
    }
}

File: learning-factory/warehouse/artifacts/evaluations/build-your-own-x/54fbf20b587133619a1f/review-v1/job_byox_review_v1_0f6667d0257bdf71d9995b689f1b4959/attempt-001/CANDIDATE/sealed/reference/parser.js
import { ParseError } from "./errors.js";
import { TokenType, tokenize } from "./lexer.js";

const BINARY_OPERATOR = Object.freeze({
  [TokenType.PLUS]: "+",
  [TokenType.MINUS]: "-",
  [TokenType.STAR]: "*",
  [TokenType.SLASH]: "/",
  [TokenType.EQUAL_EQUAL]: "==",
  [TokenType.BANG_EQUAL]: "!=",
  [TokenType.LESS]: "<",
  [TokenType.LESS_EQUAL]: "<=",
  [TokenType.GREATER]: ">",
  [TokenType.GREATER_EQUAL]: ">=",
});

class Parser {
  constructor(tokens) {
    if (!Array.isArray(tokens) || tokens.length === 0) {
      throw new TypeError("parse expects source text or a non-empty token array");
    }
    if (tokens.at(-1)?.type !== TokenType.EOF) {
      throw new ParseError("token stream must end with EOF", tokens.at(-1));
    }
    this.tokens = tokens;
    this.current = 0;
  }

  parseProgram() {
    const body = [];
    while (!this.check(TokenType.EOF)) body.push(this.statement());
    this.consume(TokenType.EOF, "expected end of input");
    return { type: "Program", body };
  }

  statement() {
    if (this.match(TokenType.LET)) return this.letStatement();
    if (this.match(TokenType.SET)) return this.setStatement();
    if (this.match(TokenType.EMIT)) return this.emitStatement();
    if (this.match(TokenType.IF)) return this.ifStatement();
    if (this.match(TokenType.WHILE)) return this.whileStatement();
    throw this.error(this.peek(), "expected a statement");
  }

  letStatement() {
    const name = this.consume(TokenType.IDENTIFIER, "expected a variable name").lexeme;
    this.consume(TokenType.EQUAL, "expected '=' after variable name");
    const initializer = this.expression();
    this.consume(TokenType.SEMICOLON, "expected ';' after expression");
    return { type: "LetStatement", name, initializer };
  }

  setStatement() {
    const name = this.consume(TokenType.IDENTIFIER, "expected a variable name").lexeme;
    this.consume(TokenType.EQUAL, "expected '=' after variable name");
    const value = this.expression();
    this.consume(TokenType.SEMICOLON, "expected ';' after expression");
    return { type: "SetStatement", name, value };
  }

  emitStatement() {
    const expression = this.expression();
    this.consume(TokenType.SEMICOLON, "expected ';' after expression");
    return { type: "EmitStatement", expression };
  }

  ifStatement() {
    const condition = this.expression();
    const consequent = this.block();
    const alternate = this.match(TokenType.ELSE) ? this.block() : null;
    return { type: "IfStatement", condition, consequent, alternate };
  }

  whileStatement() {
    const condition = this.expression();
    const body = this.block();
    return { type: "WhileStatement", condition, body };
  }

  block() {
    this.consume(TokenType.LEFT_BRACE, "expected '{'");
    const body = [];
    while (!this.check(TokenType.RIGHT_BRACE) && !this.check(TokenType.EOF)) {
      body.push(this.statement());
    }
    this.consume(TokenType.RIGHT_BRACE, "expected '}' after block");
    return { type: "BlockStatement", body };
  }

  expression() {
    return this.equality();
  }

  equality() {
    let expression = this.comparison();
    while (this.match(TokenType.EQUAL_EQUAL, TokenType.BANG_EQUAL)) {
      const operator = BINARY_OPERATOR[this.previous().type];
      const right = this.comparison();
      expression = { type: "BinaryExpression", operator, left: expression, right };
    }
    return expression;
  }

  comparison() {
    let expression = this.term();
    while (this.match(TokenType.LESS, TokenType.LESS_EQUAL, TokenType.GREATER, TokenType.GREATER_EQUAL)) {
      const operator = BINARY_OPERATOR[this.previous().type];
      const right = this.term();
      expression = { type: "BinaryExpression", operator, left: expression, right };
    }
    return expression;
  }

  term() {
    let expression = this.factor();
    while (this.match(TokenType.PLUS, TokenType.MINUS)) {
      const operator = BINARY_OPERATOR[this.previous().type];
      const right = this.factor();
      expression = { type: "BinaryExpression", operator, left: expression, right };
    }
    return expression;
  }

  factor() {
    let expression = this.unary();
    while (this.match(TokenType.STAR, TokenType.SLASH)) {
      const operator = BINARY_OPERATOR[this.previous().type];
      const right = this.unary();
      expression = { type: "BinaryExpression", operator, left: expression, right };
    }
    return expression;
  }

  unary() {
    if (this.match(TokenType.MINUS, TokenType.BANG)) {
      const operator = this.previous().type === TokenType.MINUS ? "-" : "!";
      return { type: "UnaryExpression", operator, argument: this.unary() };
    }
    return this.primary();
  }

  primary() {
    if (this.match(TokenType.NUMBER)) {
      return { type: "NumberLiteral", value: this.previous().literal };
    }
    if (this.match(TokenType.TRUE, TokenType.FALSE)) {
      return { type: "BooleanLiteral", value: this.previous().type === TokenType.TRUE };
    }
    if (this.match(TokenType.IDENTIFIER)) {
      return { type: "Identifier", name: this.previous().lexeme };
    }
    if (this.match(TokenType.LEFT_PAREN)) {
      const expression = this.expression();
      this.consume(TokenType.RIGHT_PAREN, "expected ')' after expression");
      return expression;
    }
    throw this.error(this.peek(), "expected an expression");
  }

  match(...types) {
    for (const type of types) {
      if (this.check(type)) {
        this.advance();
        return true;
      }
    }
    return false;
  }

  consume(type, message) {
    if (this.check(type)) return this.advance();
    throw this.error(this.peek(), message);
  }

  check(type) {
    return this.peek()?.type === type;
  }

  advance() {
    if (this.current < this.tokens.length) this.current += 1;
    return this.previous();
  }

  peek() {
    return this.tokens[this.current] ?? this.tokens.at(-1);
  }

  previous() {
    return this.tokens[this.current - 1];
  }

  error(token, message) {
    return new ParseError(message, token);
  }
}

/** Parse source text or the public token array returned by tokenize(). */
export function parse(sourceOrTokens) {
  const tokens = typeof sourceOrTokens === "string" ? tokenize(sourceOrTokens) : sourceOrTokens;
  return new Parser(tokens).parseProgram();
}
